crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs

```diff
@@ -39,10 +42,127 @@ impl AlwaysFixableViolation for UnnecessaryListIndexLookup {
     }
 
     fn fix_title(&self) -> String {
-        format!("Use existing item variable instead")
+        format!("Use existing variable")
     }
 }
 
+/// PLR1736
+pub(crate) fn unnecessary_list_index_lookup(checker: &mut Checker, stmt_for: &StmtFor) {
+    let Some((sequence, index_name, value_name)) =
+        enumerate_items(&stmt_for.iter, &stmt_for.target, checker.semantic())
+    else {
+        return;
+    };
+
+    let ranges = {
+        let mut visitor = SubscriptVisitor::new(sequence, index_name);
+        visitor.visit_body(&stmt_for.body);
+        visitor.visit_body(&stmt_for.orelse);
+        visitor.diagnostic_ranges
+    };
+
+    for range in ranges {
+        let mut diagnostic = Diagnostic::new(UnnecessaryListIndexLookup, range);
+        diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+            value_name.to_string(),
+            range,
+        )));
+        checker.diagnostics.push(diagnostic);
+    }
+}
+
+/// PLR1736
+pub(crate) fn unnecessary_list_index_lookup_comprehension(checker: &mut Checker, expr: &Expr) {
+    let (Expr::GeneratorExp(ast::ExprGeneratorExp {
+        elt, generators, ..
+    })
+    | Expr::DictComp(ast::ExprDictComp {
+        value: elt,
+        generators,
+        ..
+    })
+    | Expr::SetComp(ast::ExprSetComp {
+        elt, generators, ..
+    })
+    | Expr::ListComp(ast::ExprListComp {
+        elt, generators, ..
+    })) = expr
+    else {
+        return;
+    };
+
+    for comp in generators {
+        let Some((sequence, index_name, value_name)) =
+            enumerate_items(&comp.iter, &comp.target, checker.semantic())
+        else {
+            return;
+        };
+
+        let ranges = {
+            let mut visitor = SubscriptVisitor::new(sequence, index_name);
+            visitor.visit_expr(&elt.as_ref());
+            visitor.diagnostic_ranges
+        };
+
+        for range in ranges {
+            let mut diagnostic = Diagnostic::new(UnnecessaryListIndexLookup, range);
+            diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+                value_name.to_string(),
+                range,
+            )));
+            checker.diagnostics.push(diagnostic);
+        }
+    }
+}
```

Comment:
I prefer to put the rule implementations up top, and helper functions at the bottom. That way, each file reads as (1) diagnostic definition, then (2) rule body.